mno_analysis_tools/compute_messages_per_period.py

```diff
@@ -0,0 +1,114 @@
+    parser.add_argument("time_frame", metavar="time-frame", type=lambda s: datetime.strptime(s, '%H:%M:%S'),
+                        help="The time frame (HH:MM:SS) to generate dates in intervals between the start and end date")
+
+    args = parser.parse_args()
+
+    raw_messages_file_path = args.raw_messages_file_path
+    computed_messages_per_period_file_path = args.computed_messages_per_period_file_path
+    message_difference_file_path = args.message_difference_file_path
+    target_operator = args.target_operator
+    target_message_direction = args.target_message_direction
+    start_date = args.start_date
+    end_date = args.end_date
+    time_frame = args.time_frame
+
+    with open(raw_messages_file_path, mode="r") as f:
+        log.info("Loading messages from {raw_messages_file_path}...")
+        input = json.load(f)
+        messages = [Message.deserialize(val) for val in input]
+        log.info(f"Loaded {len(messages)} messages")
+
+    # Filter messages based on the target operator and target direction of the message
+    log.info(f"Filter messages based on operator {target_operator} and, "
+             "message direction as {target_message_direction}")
+    log.info(f"{len(messages)} messages before filtering")
+    filtered_messages = []
+    for msg in messages:
+        if msg.urn.startswith("tel:"):
+            operator = PhoneCleaner.clean_operator(msg.urn.split(":")[1])
+        else:
+            operator = msg.urn.split(":")[0]
+        if operator == target_operator and msg.direction == target_message_direction:
+            log.info(f"Filtering messages...")
+            filtered_messages.append(msg)
+    log.info(f"{len(filtered_messages)} messages after filtering")
+
+    time_interval = timedelta(hours=time_frame.hour,
+                              minutes=time_frame.minute, seconds=time_frame.second)
+
+    date_time_bounds = date_time_range(start_date, end_date, time_interval)
+
+    # Compute number of messages between two datetime bounds i.e `PreviousMessageTimestamp` and
+    # `NextMessageTimestamp` to get number of mesages per period and relate each quantity
+    #  with the operator and the message direction.
+    messages_per_period = []
+    for index in range(len(date_time_bounds) - 1):
+        next_index = index + 1
+
+        messages_this_period = 0
+        for msg in filtered_messages:
+            if date_time_bounds[index] <= msg.sent_on < date_time_bounds[next_index]:
+                messages_this_period += 1
+
+        messages_per_period.append({
+            "Operator": target_operator,
+            "MessageDirection": target_message_direction,
+            "PeriodStart": date_time_bounds[index].isoformat(),
+            "PeriodEnd": date_time_bounds[next_index].isoformat(),
+            "NumberOfMessages": messages_this_period
+        })
+
+    message_difference_per_period = []
```

Comment:
Does this need to be separate? Or could it go in the messages_per_period as a "Delta"? Or computed with a switch in Sheets when you generate the graphs?

Reply:
I had thought about this before.. I think it should be separate. If we're going to compute this when generating the graphs, the script will slow down the browser considering we're going to fetch all the data from rapid pro for Imaqal.  I also could not pass it as a delta as message difference per period is dependent on messages per period